src/iotjs_binding_helper.c

```diff
@@ -28,24 +28,25 @@ void iotjs_uncaught_exception(jerry_value_t jexception) {
   IOTJS_ASSERT(jerry_value_is_function(jonuncaughtexception));
 
   iotjs_jargs_t args = iotjs_jargs_create(1);
+  jerry_value_clear_error_flag(&jexception);
   iotjs_jargs_append_jval(&args, jexception);
 
-  bool throws;
-  jerry_value_t jres =
-      iotjs_jhelper_call(jonuncaughtexception, process, &args, &throws);
+  jerry_value_t jres = iotjs_jhelper_call(jonuncaughtexception, process, &args);
 
   iotjs_jargs_destroy(&args);
-  jerry_release_value(jres);
   jerry_release_value(jonuncaughtexception);
 
-  if (throws) {
+  if (jerry_value_has_error_flag(jres)) {
     iotjs_environment_t* env = iotjs_environment_get();
 
     if (!iotjs_environment_is_exiting(env)) {
       iotjs_set_process_exitcode(2);
       iotjs_environment_go_state_exiting(env);
     }
   }
+
+  jerry_release_value(jexception);
```

Comment:
@LaszloLango Before this PR, it was `iotjs_uncaught_exception(jexception)` caller's role to release jexception. After this PR, the 'iotjs_uncaught_exception` will always release `jexception` parameter internally. So the user have to acquire the jexception before call. (jerry_acquire_value is added in that reason) Usually I don't expect my parameter will be released implicitly by calling some function. Could you explain why do you want to change the behavior? 

I prefer not to release the parameter, if there is compelling reason. Because if some functions releases the parameters, while others not, it will make us confused. And it would be better to emphasize the function user. "It will release your parameter." by comment or by using descriptive function name. (In that case, I would choose latter. Because users don't read function comment, and comment is to be out of date.)

Reply:
AFAIK `iotjs_uncaught_exception` is an internal function an shouldn't be used by a user. The user must register a callback, but the IoT.js must call the callback. So I do not understand your concerns.

Technical details:
When you clear the error flag it will give back a new `jerrry_value_t` and if there where no other reference, than the original value is freed. So the `jmain` will contain an invalid (freed) value at the call site of `iotjs_uncaught_exception`  (`iotjs_run`). This is why we use `jerry_acquire_value` before calling `iotjs_uncaught_exception`.

The other way would be to give back the return value of `jerry_value_clear_error_flag`.

`jerry_value_t iotjs_uncaught_exception(jerry_value_t jexception)`
or
`void iotjs_uncaught_exception(jerry_value_t* jexception)`

Reply:
@LaszloLango 
When I mentioned a user, I mean the iotjs core developer like me. There were many changes in iotjs by many developers. I found several different style and convention. So I think it is important to keep the code consistent, and readable. I would like to give effort in our internal code, not only for external api.

This change makes it difficult to read. Imagine my string is release after I call function `foo(myStr)`, and `foo` don't provide comment, and cannot guess it will release my string from `foo`.

Could you please answer to my original question?
Still I cannot find what is the benefit of moving release into  `iotjs_uncaught_exception`.

Reply:
Sorry I thought I could answer that question. :) The JerryScript API is intended to use in this way. What about the suggested alternatives?

>The other way would be to give back the return value of jerry_value_clear_error_flag.
>
>jerry_value_t iotjs_uncaught_exception(jerry_value_t jexception)
>or
>void iotjs_uncaught_exception(jerry_value_t* jexception)

Would it be more readable for you?

Reply:
@LaszloLango Hmm, my confusion comes from the modified semantics of `jerry_value_clear_error_flag`. It introduced several issues(https://github.com/jerryscript-project/jerryscript/issues/2080, https://github.com/jerryscript-project/jerryscript/pull/2131) and added an API(https://github.com/jerryscript-project/jerryscript/pull/2096). I need to take a closer look for this API and implementation to review this PR.

Reply:
@LaszloLango The comment `jerry_get_value_without_error_flag` (https://github.com/jerryscript-project/jerryscript/issues/2080#issuecomment-343445688) was very helpful. What about using `jerry_get_value_without_error_flag` on caller of `iotjs_uncaught_exception` like code below?

```C
jerry_value_t errval = jerry_get_value_without_error_flag(jres);
iotjs_uncaught_exception(errval);
jerry_release_value(errval);
```

Reply:
If I understand you correctly and you wish to keep "errval" alive, then iotjs_uncaught_exception() could do the reference increase itself. So you can do the usual "call free for any jres".

```
val = something()
if (jerry_is_value_error(val))
  iotjs_uncaught_exception(val);
jerry_release_value(val);
```

Reply:
@zherczeg Right.

I feel odds when I read the code because:

`err_val` (which is passed to `iotjs_uncaught_exception` by value) is released implicitly. Therefore we have to increase reference count for `err_val` before calling `iotjs_uncaught_exception`. 

I think there are two approach to explain the code more clearly.

(1) increase `err_val` in `iotjs_uncaught_exception()` (@zherczeg's https://github.com/Samsung/iotjs/pull/1326#discussion_r160096745)
(2) use `jerry_value_get_without_error_flag` for getting a copy of err_val without error flag, and passing it to `iotjs_uncaught_exception` (my https://github.com/Samsung/iotjs/pull/1326#discussion_r159368404)

Either works for me.

@LaszloLango I don't saying your implementation is wrong. But I would like to write the code differently so that others who are not accustomed for jerry could understand more easily. Because most don't know the details that you explained on https://github.com/Samsung/iotjs/pull/1326#discussion_r156869477

Reply:
Sorry for the late answer. I missed your previous comment. I'll update the PR as you suggested.